app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt

```diff
@@ -30,6 +36,130 @@ constructor(
     checkUserLoginStatus()
   }
 
+  fun onSearchQueryChanged(
+      query: String,
+      state: MutableStateFlow<EventsOverviewUiState> = _uiState
+  ) {
+    state.value = state.value.copy(searchQuery = query)
+    filterEvents()
+  }
+
+  fun onSearchActiveChanged(
+      isSearchActive: Boolean,
+      state: MutableStateFlow<EventsOverviewUiState> = _uiState
+  ) {
+    state.value = state.value.copy(isFilterDialogOpen = false)
+    state.value = state.value.copy(isSearchActive = isSearchActive)
+  }
+
+  fun onFilterDialogOpen(state: MutableStateFlow<EventsOverviewUiState> = _uiState) {
+    state.value = state.value.copy(isFilterDialogOpen = !state.value.isFilterDialogOpen)
+  }
+
+  fun onRadiusQueryChanged(
+      radius: String,
+      state: MutableStateFlow<EventsOverviewUiState> = _uiState
+  ) {
+    state.value = state.value.copy(radiusQuery = radius)
+  }
+
+  fun onFreeSwitchChanged(state: MutableStateFlow<EventsOverviewUiState> = _uiState) {
+    state.value = state.value.copy(isFreeSwitchOn = !state.value.isFreeSwitchOn)
+  }
+
+  fun onFilterApply(state: MutableStateFlow<EventsOverviewUiState> = _uiState) {
+    state.value = state.value.copy(isFilterActive = true)
+    filterEvents()
+  }
+
+  fun filterEvents() {
```

Comment:
**Suggestion:** The filterEvents() method could be improved by combining the multiple filtering operations into a single pass over the event list. Currently, the method filters the list separately for search queries, radius, free events, and categories, which is not efficient.

Additionally, the decision logic to determine the list based on the Tab is split between the start and end of the method. It would be cleaner to handle this selection only once.

Reply:
I have made filterEvents more efficient by combining all filters into one. 

Regarding the tab split logic, it is not possible to do this only once without making the code even less efficient. i.e. this is the most efficient way to do this. Other means would require repeating logic.